internal/core/command/init.go

```diff
@@ -76,107 +82,119 @@ func Retry(useConsul bool, useProfile string, timeout int, wait *sync.WaitGroup,
+		ServicePort:     conf.Service.Port,
+		ServiceProtocol: conf.Service.Protocol,
+		CheckInterval:   conf.Service.CheckInterval,
+		CheckRoute:      clients.ApiPingRoute,
+		Stem:            internal.ConfigRegistryStem,
+	}
 
+	registryClient, err = factory.NewRegistryClient(registryConfig)
 	if err != nil {
-		return conf, fmt.Errorf("connection to Consul could not be made: %v", err.Error())
+		return fmt.Errorf("connection to Registry could not be made: %v", err.Error())
 	}
-	// Update configuration data from Consul
-	updateCh := make(chan interface{})
-	errCh := make(chan error)
-	dec := consulclient.NewConsulDecoder(conf.Registry)
-	dec.Target = &ConfigurationStruct{}
-	dec.Prefix = internal.ConfigRegistryStem + internal.CoreCommandServiceKey
-	dec.ErrCh = errCh
-	dec.UpdateCh = updateCh
-
-	defer dec.Close()
-	defer close(updateCh)
-	defer close(errCh)
-	go dec.Run()
-
-	select {
-	case <-time.After(2 * time.Second):
-		err = errors.New("timeout loading config from registry")
-	case ex := <-errCh:
-		err = errors.New(ex.Error())
-	case raw := <-updateCh:
-		actual, ok := raw.(*ConfigurationStruct)
-		if !ok {
-			return conf, errors.New("type check failed")
-		}
-		conf = actual
-		//Check that information was successfully read from Consul
-		if conf.Service.Port == 0 {
-			return nil, errors.New("error reading from Consul")
-		}
+
+	// Check if registry service is running
+	if !registryClient.IsAlive() {
+		return fmt.Errorf("registry is not available")
 	}
 
-	return conf, err
+	// Register the service with Registry
+	err = registryClient.Register()
+	if err != nil {
+		return fmt.Errorf("could not register service with Registry: %v", err.Error())
+	}
+
+	return nil
 }
 
 func listenForConfigChanges() {
-	errCh := make(chan error)
-	dec := consulclient.NewConsulDecoder(Configuration.Registry)
-	dec.Target = &WritableInfo{}
-	dec.Prefix = internal.ConfigRegistryStem + internal.CoreCommandServiceKey + internal.WritableKey
-	dec.ErrCh = errCh
-	dec.UpdateCh = chConfig
+	if registryClient == nil {
+		LoggingClient.Error("listenForConfigChanges() registry client not set")
+		return
+	}
+
+	registryClient.WatchForChanges(updateChannel, errChannel, &WritableInfo{}, internal.WritableKey)
 
-	defer dec.Close()
-	defer close(errCh)
+	signalChan := make(chan os.Signal)
+	signal.Notify(signalChan, syscall.SIGINT, syscall.SIGTERM)
 
-	go dec.Run()
 	for {
 		select {
-		case ex := <-errCh:
+		case <-signalChan:
+			// Quietly and gracefully stop when SIGINT/SIGTERM received
+			return
+		case ex := <-errChannel:
 			LoggingClient.Error(ex.Error())
-		case raw, ok := <-chConfig:
+		case raw, ok := <-updateChannel:
 			if ok {
 				actual, ok := raw.(*WritableInfo)
 				if !ok {
 					LoggingClient.Error("listenForConfigChanges() type check failed")
 				}
 				Configuration.Writable = *actual
+				LoggingClient.Info("Writeable configuration has been updated. Setting log level to " + Configuration.Writable.LogLevel)
```

Comment:
I think it is very useful in know if you config us being update when changed in the Registry.

Reply:
OK, but the LogLevel is only one setting within the Writable section and it may not be the one that actually changed. If we want to write a log entry like this, might be better to put it in the LoggingClient.SetLogLevel() function where the old value can be compared to the new and, if different, write the entry.

Reply:
I just pick logging level because we are doing something with it. The main goal was to indicate that Writeable configuration was updated from the Registry. How about a more generic message like "Writeable configuration has been updated from the Registry".

Reply:
That would be fine